Documentation/specs/corehost.md

```diff
@@ -0,0 +1,112 @@
+# corehost runtime/assembly resolution
+
+The shared host locates assemblies and native libraries using a combination of: Servicing Index, Files in the application folder (aka "app-local") and files from package caches.
+
+## Definitions and Formats
+
+### Servicing Index
+
+The servicing index is loaded when the `DOTNET_SERVICING` environment variable is non-empty. When this variable is non-empty, it points to a directory that will be the **Servicing Root**.
+
+An index file is located at the path defined by `$DOTNET_SERVICING/servicing_index.txt`. In this file are a series of lines of one of the following formats:
+
+```
+# Lines starting with a '#' and blank lines are ignored.
+
+# Identifies an asset from a NuGet Package that has been serviced.
+package|[Package ID]|[Package Version]|[Original Asset Relative Path]=[New Asset Path, relative to Servicing Root]
+package|System.Threading.Thread|1.2.3.4|lib/dotnet5.4/System.Threading.Thread.dll=patches/abc123/System.Threading.Thread.dll
+
+# TBD: Host/Runtime servicing entries.
+```
+
+Paths in this file are **always** specified using `/`, even on Windows. They must be converted to platform-specific directory separators.
+
+This index is loaded when needed during the Resolution Process (see below).
+
+### Dependencies File
+
+The `.deps` file is a file located alongside the native host. It is stored in a CSV format, with header, of the following format:
+
+```
+LibraryType,LibraryName,LibraryVersion,LibraryHash,AssetType,AssetName,AssetRelativePath
+"Package","Microsoft.CodeAnalysis.Common","1.2.0-beta-20151117-04","sha512-4O3f1iRswbWDz70AGTk0tBA1ivslJUuq37FSx4d0+ZuoBFxNO4oY7ReJgndOJYefaxeTMsCVHSVHJMYOJJphtQ==","runtime","Microsoft.CodeAnalysis","lib\\portable-net45+win8\\Microsoft.CodeAnalysis.dll",
+```
+
+Fields:
+
+* `LibraryType` -> The type of Library that provides the Asset. Currently always `Package` (other Library types include Projects and Framework Assemblies but these aren't relevant to `corehost`)
+* `LibraryName` -> The name of the Library that provides the Asset. For a Package, this is the Package ID
+* `LibraryVersion` -> The version of the Library that provides the Asset. For a Package, this is the Package Version
+* `LibraryHash` -> The hash of the Library that provides the Asset. For a Package, this is a hash of the `.nupkg` file. The format is `[algorithm]-[base64-encoded hash]` where `[algorithm]` must always be `sha512` (in the current version)
+* `AssetType` -> The type of the Asset within the Library. `corehost` recognizes two types: `runtime` and `native`, representing Managed Assemblies and Native Libraries respectively.
```

Comment:
Yeah, the naming of assets in the lock file is: `compile` (for things passed to csc), `runtime` (for things used at runtime) and `native` (because we need to separate native and managed libraries at runtime). We should keep the `.deps` file in sync with that.

I do agree that `managed` would probably be a better name, but the lock file has shipped in RTM products so it's probably too late to change that :). Maybe just consider `runtime` as referring to the CoreCLR Runtime :stuck_out_tongue: